### PR TITLE
Qgis dimensions

### PR DESCRIPTION
**Pull request by arnaud-morvan**
Improve QGIS compatibility : 
* Add field property to dimensions.
* Add version_wms to ogc_server.

### CODE REVIEW

File: c2cgeoportal/models.py [hunk 1]
```diff
@@ -1021,10 +1022,12 @@ class Dimension(Base):
             cascade="save-update,merge,delete,delete-orphan",
         ),
     )
+    field = Column(Unicode, label=_(u"Field"))
```
Comment (sbrunner) [inline c2cgeoportal/models.py]: Why not put it just around the description?

File: c2cgeoportal/views/entry.py [hunk 2]
```diff
@@ -73,15 +73,17 @@
 
 class DimensionInformation:
 
-    URL_PART_RE = re.compile("[a-zA-Z0-9_\-\+~\.]+$")
+    URL_PART_RE = re.compile("[a-zA-Z0-9_\-\+~\.]*$")
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: Why this change?

File: c2cgeoportal/models.py [hunk 3]
```diff
@@ -710,6 +710,7 @@ class OGCServer(Base):
     name = Column(Unicode, label=_(u"Name"), nullable=False, unique=True)
     description = Column(Unicode, label=_(u"Description"))
     url = Column(Unicode, label=_(u"Base URL"), nullable=False)
+    version_wms = Column(Unicode, label=_(u"WMS version"))
```
Comment (sbrunner) [inline c2cgeoportal/models.py]: Default to 1.1.1?

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+revision = '72eaad5daf34'
+down_revision = '809650bd04c3'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    schema = context.get_context().config.get_main_option('schema')
+
+    op.add_column('ogc_server', Column('version_wms', Unicode), schema=schema)
```
Comment (sbrunner) [inline c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py]: Default to 1.1.1?

File: c2cgeoportal/views/entry.py [hunk 5]
```diff
@@ -269,16 +280,20 @@ def _wms_getcap_cached(self, ogc_server, role_id):
...
-            wms = WebMapService(None, xml=content)
-        except:  # pragma: no cover
+            wms = WebMapService(None, xml=content, version=version)
+        except Exception as e:  # pragma: no cover
             error = _(
                 "WARNING! an error occurred while trying to "
                 "read the mapfile and recover the themes."
             )
-            error = u"{0!s}\nURL: {1!s}\n{2!s}".format(error, url, content.encode("utf-8"))
+            error = u"{0!s}\nURL: {1!s}".format(error, url)
```
Comment (sbrunner) [inline c2cgeoportal/views/entry.py]: Why removing the content?

File: c2cgeoportal/views/entry.py (hunk 2 above)
Comment (arnaud-morvan) [reply c2cgeoportal/views/entry.py]: This is a backport of merged commit in 2.3 (support NULL dimensions), because I need it to work with ADL.

Comment (sbrunner) [reply c2cgeoportal/views/entry.py]: OK :-)

File: c2cgeoportal/models.py (hunk 1 above)
Comment (arnaud-morvan) [reply c2cgeoportal/models.py]: Because I like field to appear after layer in admin UI

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py (hunk 4 above)
Comment (arnaud-morvan) [reply c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py]: This is done in Entry class `version_wms or '1.1.1'`

File: c2cgeoportal/views/entry.py (hunk 5 above)
Comment (arnaud-morvan) [reply c2cgeoportal/views/entry.py]: I removed content because it's hard to find error as content is very long.
Note that you have the url, easy to get content if needed.

File: c2cgeoportal/models.py (hunk 1 above)
Comment (sbrunner) [reply c2cgeoportal/models.py]: The field order can be change like this:
https://github.com/arnaud-morvan/c2cgeoportal/blob/f4f07948167539be05f9291838a13ffbfbeee05e/c2cgeoportal/forms.py#L704-L709

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py (hunk 4 above)
Comment (sbrunner) [reply c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py]: For me the default value should be visible in the admin interface...

File: c2cgeoportal/views/entry.py (hunk 5 above)
Comment (sbrunner) [reply c2cgeoportal/views/entry.py]: But many times the error is in the content, or related to the content...

Comment (arnaud-morvan) [reply c2cgeoportal/views/entry.py]: I note that it may be useful if status is 200 but response is a service exception.
In case response is a capabilities, like in my use case it's a PITA.
I did not plan to commit, I will revert this point.

File: c2cgeoportal/models.py (hunk 1 above)
Comment (arnaud-morvan) [reply c2cgeoportal/models.py]: And it's another thing in 2.3

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py (hunk 4 above)
Comment (arnaud-morvan) [reply c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py]: I also like the idea that we leave the choice to the app if the user does not want to force.
(=> set this only if you know what you are doing)
What about if you want to change the default value in future version ?

Comment (arnaud-morvan) [reply c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py]: In 2.3 we can add a description for the field (default to 1.1.1)

File: c2cgeoportal/models.py (hunk 1 above)
Comment (sbrunner) [reply c2cgeoportal/models.py]: Yes but as I know you didn't move the fields in the model to change the order, or yes?

File: c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py (hunk 4 above)
Comment (sbrunner) [reply c2cgeoportal/scaffolds/update/CONST_alembic/main/versions/72eaad5daf34_add_field_ogc_server_version_wms.py]: OK for the description :-)

File: c2cgeoportal/models.py (hunk 1 above)
Comment (arnaud-morvan) [reply c2cgeoportal/models.py]: As I remember, we did not change any order, we show fields in sqlalchemy model order.
Except for some relations, we create schema manually and insert them imperatively at the right place.
Concerning dimensions, as they are embedded in layer form, layer field should be excluded.
So you're right, we could put field before layer.

Comment (sbrunner) [reply c2cgeoportal/models.py]: cool :-)

**Comment by arnaud-morvan** [comment]
Replaced by #4112
